{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "from statsmodels.stats.meta_analysis import combine_effects\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "os.chdir('/project/ssverma_shared/projects/Endometriosis/Endo_Subtyping_Heterogeneity/Meta_Analysis/')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001B[33mBuilding DAG of jobs...\u001B[0m\r\n",
      "\u001B[33mNothing to be done (all requested files are present and up to date).\u001B[0m\r\n"
     ]
    }
   ],
   "source": [
    "! snakemake -n all --quiet"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [],
   "source": [
    "phenos = ['endometriosis',\n",
    "          'cluster_vs_controls_0',\n",
    "          'cluster_vs_controls_1',\n",
    "          'cluster_vs_controls_2',\n",
    "          'cluster_vs_controls_3',\n",
    "          'cluster_vs_controls_4']"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [],
   "source": [
    "pheno_ss = {p: [] for p in phenos}\n",
    "pheno_ss_AFR_EUR = {p: [] for p in phenos}\n",
    "pheno_ss_EUR = {p: [] for p in phenos}\n",
    "\n",
    "for d in ['eMERGE', 'UKBB', 'PMBB']:\n",
    "    for a in ['AFR', 'EUR', 'ASIAN']:\n",
    "        if d == 'eMERGE' and a == 'ASIAN':\n",
    "            continue\n",
    "        for f in os.listdir(f'{d}_{a}/Sumstats/'):\n",
    "            file_pheno = f.split('.')[0]\n",
    "            pheno_ss[file_pheno].append(f'{d}_{a}/Sumstats/{f}')\n",
    "            if a != 'ASIAN':\n",
    "                pheno_ss_AFR_EUR[file_pheno].append(f'{d}_{a}/Sumstats/{f}')\n",
    "            if a == 'EUR':\n",
    "                pheno_ss_EUR[file_pheno].append(f'{d}_{a}/Sumstats/{f}')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [],
   "source": [
    "eMERGE_id_map = pd.read_table('SNP_Lists/eMERGE_to_PMBB_ID_map.txt', index_col=['ID'])['PMBB_ID']\n",
    "UKBB_id_map = pd.read_table('SNP_Lists/UKBB_to_PMBB_ID_map.txt', index_col=['ID'])['PMBB_ID']"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "endometriosis\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/endometriosis.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.endometriosis.txt Meta_Input/eMERGE_EUR.endometriosis.txt Meta_Input/UKBB_AFR.endometriosis.txt Meta_Input/UKBB_EUR.endometriosis.txt Meta_Input/UKBB_ASIAN.endometriosis.txt Meta_Input/PMBB_AFR.endometriosis.txt Meta_Input/PMBB_EUR.endometriosis.txt Meta_Input/PMBB_ASIAN.endometriosis.txt\r\n",
      "  --out Meta_Output/endometriosis.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/endometriosis.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/endometriosis.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.endometriosis.txt Meta_Input/eMERGE_EUR.endometriosis.txt Meta_Input/UKBB_AFR.endometriosis.txt Meta_Input/UKBB_EUR.endometriosis.txt Meta_Input/PMBB_AFR.endometriosis.txt Meta_Input/PMBB_EUR.endometriosis.txt\r\n",
      "  --out Meta_Output/endometriosis.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/endometriosis.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/endometriosis.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.endometriosis.txt Meta_Input/UKBB_EUR.endometriosis.txt Meta_Input/PMBB_EUR.endometriosis.txt\r\n",
      "  --out Meta_Output/endometriosis.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/endometriosis.EUR.PLINK.meta .\r\n",
      "cluster_vs_controls_0\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_0.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/UKBB_EUR.cluster_vs_controls_0.txt Meta_Input/PMBB_AFR.cluster_vs_controls_0.txt Meta_Input/PMBB_EUR.cluster_vs_controls_0.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_0.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 789 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_0.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_0.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/UKBB_EUR.cluster_vs_controls_0.txt Meta_Input/PMBB_AFR.cluster_vs_controls_0.txt Meta_Input/PMBB_EUR.cluster_vs_controls_0.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_0.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 789 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_0.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_0.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/UKBB_EUR.cluster_vs_controls_0.txt Meta_Input/PMBB_EUR.cluster_vs_controls_0.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_0.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 789 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_0.EUR.PLINK.meta .\r\n",
      "cluster_vs_controls_1\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_1.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_1.txt Meta_Input/UKBB_AFR.cluster_vs_controls_1.txt Meta_Input/UKBB_EUR.cluster_vs_controls_1.txt Meta_Input/UKBB_ASIAN.cluster_vs_controls_1.txt Meta_Input/PMBB_AFR.cluster_vs_controls_1.txt Meta_Input/PMBB_EUR.cluster_vs_controls_1.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_1.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_1.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_1.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_1.txt Meta_Input/UKBB_AFR.cluster_vs_controls_1.txt Meta_Input/UKBB_EUR.cluster_vs_controls_1.txt Meta_Input/PMBB_AFR.cluster_vs_controls_1.txt Meta_Input/PMBB_EUR.cluster_vs_controls_1.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_1.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_1.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_1.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_1.txt Meta_Input/UKBB_EUR.cluster_vs_controls_1.txt Meta_Input/PMBB_EUR.cluster_vs_controls_1.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_1.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_1.EUR.PLINK.meta .\r\n",
      "cluster_vs_controls_2\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_2.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.cluster_vs_controls_2.txt Meta_Input/eMERGE_EUR.cluster_vs_controls_2.txt Meta_Input/UKBB_EUR.cluster_vs_controls_2.txt Meta_Input/UKBB_ASIAN.cluster_vs_controls_2.txt Meta_Input/PMBB_AFR.cluster_vs_controls_2.txt Meta_Input/PMBB_EUR.cluster_vs_controls_2.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_2.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_2.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_2.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.cluster_vs_controls_2.txt Meta_Input/eMERGE_EUR.cluster_vs_controls_2.txt Meta_Input/UKBB_EUR.cluster_vs_controls_2.txt Meta_Input/PMBB_AFR.cluster_vs_controls_2.txt Meta_Input/PMBB_EUR.cluster_vs_controls_2.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_2.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_2.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_2.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_2.txt Meta_Input/UKBB_EUR.cluster_vs_controls_2.txt Meta_Input/PMBB_EUR.cluster_vs_controls_2.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_2.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_2.EUR.PLINK.meta .\r\n",
      "cluster_vs_controls_3\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_3.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_3.txt Meta_Input/UKBB_EUR.cluster_vs_controls_3.txt Meta_Input/PMBB_AFR.cluster_vs_controls_3.txt Meta_Input/PMBB_EUR.cluster_vs_controls_3.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_3.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_3.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_3.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_3.txt Meta_Input/UKBB_EUR.cluster_vs_controls_3.txt Meta_Input/PMBB_AFR.cluster_vs_controls_3.txt Meta_Input/PMBB_EUR.cluster_vs_controls_3.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_3.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_3.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_3.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_3.txt Meta_Input/UKBB_EUR.cluster_vs_controls_3.txt Meta_Input/PMBB_EUR.cluster_vs_controls_3.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_3.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_3.EUR.PLINK.meta .\r\n",
      "cluster_vs_controls_4\n",
      "\t ALL\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_4.ALL.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.cluster_vs_controls_4.txt Meta_Input/eMERGE_EUR.cluster_vs_controls_4.txt Meta_Input/UKBB_EUR.cluster_vs_controls_4.txt Meta_Input/PMBB_AFR.cluster_vs_controls_4.txt Meta_Input/PMBB_EUR.cluster_vs_controls_4.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_4.ALL.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_4.ALL.PLINK.meta .\r\n",
      "\t AFR_EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_4.AFR_EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_AFR.cluster_vs_controls_4.txt Meta_Input/eMERGE_EUR.cluster_vs_controls_4.txt Meta_Input/UKBB_EUR.cluster_vs_controls_4.txt Meta_Input/PMBB_AFR.cluster_vs_controls_4.txt Meta_Input/PMBB_EUR.cluster_vs_controls_4.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_4.AFR_EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 748 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_4.AFR_EUR.PLINK.meta .\r\n",
      "\t EUR\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\r\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\r\n",
      "Logging to Meta_Output/cluster_vs_controls_4.EUR.PLINK.log.\r\n",
      "Options in effect:\r\n",
      "  --meta-analysis Meta_Input/eMERGE_EUR.cluster_vs_controls_4.txt Meta_Input/UKBB_EUR.cluster_vs_controls_4.txt Meta_Input/PMBB_EUR.cluster_vs_controls_4.txt\r\n",
      "  --out Meta_Output/cluster_vs_controls_4.EUR.PLINK\r\n",
      "\r\n",
      "128235 MB RAM detected; reserving 64117 MB for main workspace.\r\n",
      "--meta-analysis: 744 variants processed; results written to\r\n",
      "Meta_Output/cluster_vs_controls_4.EUR.PLINK.meta .\r\n"
     ]
    }
   ],
   "source": [
    "os.makedirs('Meta_Input/', exist_ok=True)\n",
    "\n",
    "for p in phenos:\n",
    "    print(p)\n",
    "    effs = pd.DataFrame()\n",
    "    errs = pd.DataFrame()\n",
    "    alleles = pd.DataFrame()\n",
    "\n",
    "    # Load data\n",
    "    for ss_file in pheno_ss[p]:\n",
    "        ss_dataset = ss_file.split('_')[0]\n",
    "        ss_cohort = ss_file.split('/')[0]\n",
    "        temp = pd.read_table(ss_file, index_col='ID', nrows=None)\n",
    "        temp = temp[temp['AF_Allele2'] >= 0.01]\n",
    "        temp = temp[['BETA', 'Allele2', 'SE']].dropna()\n",
    "\n",
    "        if ss_dataset == 'eMERGE':\n",
    "            temp = temp[temp.index.isin(eMERGE_id_map.index)]\n",
    "            temp = temp.rename(index=eMERGE_id_map)\n",
    "        if ss_dataset == 'UKBB':\n",
    "            temp = temp[temp.index.isin(UKBB_id_map.index)]\n",
    "            temp = temp.rename(index=UKBB_id_map)\n",
    "\n",
    "        effs[ss_cohort] = temp['BETA']\n",
    "        errs[ss_cohort] = temp['SE']\n",
    "        alleles[ss_cohort] = temp['Allele2']\n",
    "\n",
    "    # Now we have data to subset for meta-analysis\n",
    "    effs = effs[effs.count(axis=1) > 1]\n",
    "    errs = errs.loc[effs.index]\n",
    "    alleles = alleles.loc[effs.index]\n",
    "\n",
    "    use_allele = alleles.mode(axis=1)[0]\n",
    "    use_allele_2D = np.broadcast_to(use_allele.T, alleles.T.shape).T\n",
    "    flip_coords = np.where(alleles.values != use_allele_2D)\n",
    "    effs.values[flip_coords] = -effs.values[flip_coords]\n",
    "\n",
    "    for cohort, effect_col in effs.iteritems():\n",
    "        table = pd.concat([effect_col, errs[cohort], use_allele], axis=1)\n",
    "        table.columns = ['BETA', 'SE', 'A1']\n",
    "        table.index.name = 'SNP'\n",
    "        table['OR'] = np.exp(table['BETA'])\n",
    "        table = table.dropna().reset_index()\n",
    "        table[['CHR', 'BP']] = table['SNP'].str.replace('chr', '').str.split(':', expand=True)[[0, 1]]\n",
    "        table[['CHR', 'BP', 'SNP', 'OR', 'SE', 'A1']].to_csv(f'Meta_Input/{cohort}.{p}.txt', sep='\\t', index=False)\n",
    "\n",
    "    # Use different combinations of groups\n",
    "    for group, ss_dict in zip(['ALL', 'AFR_EUR', 'EUR'], [pheno_ss, pheno_ss_AFR_EUR, pheno_ss_EUR]):\n",
    "        print('\\t', group)\n",
    "        ss_group = ss_dict[p]\n",
    "        cohorts = [f.split('/')[0] for f in ss_group]\n",
    "\n",
    "        plink_cmd_parts = ['plink', '--meta-analysis']\n",
    "        plink_cmd_parts.extend([f'Meta_Input/{cohort}.{p}.txt' for cohort in cohorts])\n",
    "        plink_cmd_parts.extend(['--out', f'Meta_Output/{p}.{group}.PLINK'])\n",
    "        cmd = ' '.join(plink_cmd_parts)\n",
    "\n",
    "        ! module load plink/1.90Beta6.18; {cmd}\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "endometriosis\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n",
      "cluster_vs_controls_0\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n",
      "cluster_vs_controls_1\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n",
      "cluster_vs_controls_2\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n",
      "cluster_vs_controls_3\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n",
      "cluster_vs_controls_4\n",
      "\t ALL\n",
      "\t AFR_EUR\n",
      "\t EUR\n"
     ]
    }
   ],
   "source": [
    "os.makedirs('Meta_Output/', exist_ok=True)\n",
    "\n",
    "for p in phenos:\n",
    "    print(p)\n",
    "    effs = pd.DataFrame()\n",
    "    vars = pd.DataFrame()\n",
    "    alleles = pd.DataFrame()\n",
    "\n",
    "    # Load data\n",
    "    for ss_file in pheno_ss[p]:\n",
    "        ss_dataset = ss_file.split('_')[0]\n",
    "        ss_cohort = ss_file.split('/')[0]\n",
    "        temp = pd.read_table(ss_file, index_col='ID', nrows=None)\n",
    "        temp['N'] = temp['N_case'] + temp['N_ctrl']\n",
    "        temp['SD'] = temp['SE'] * np.sqrt(temp['N'])\n",
    "        temp['VAR'] = temp['SD'] ** 2\n",
    "        temp = temp[temp['AF_Allele2'] >= 0.01]\n",
    "        temp = temp[['BETA', 'VAR', 'Allele2']].dropna()\n",
    "        if ss_dataset == 'eMERGE':\n",
    "            temp = temp[temp.index.isin(eMERGE_id_map.index)]\n",
    "            temp = temp.rename(index=eMERGE_id_map)\n",
    "        if ss_dataset == 'UKBB':\n",
    "            temp = temp[temp.index.isin(UKBB_id_map.index)]\n",
    "            temp = temp.rename(index=UKBB_id_map)\n",
    "\n",
    "        effs[ss_cohort] = temp['BETA']\n",
    "        vars[ss_cohort] = temp['VAR']\n",
    "        alleles[ss_cohort] = temp['Allele2']\n",
    "\n",
    "    # Now we have data to subset for meta-analysis\n",
    "    effs = effs[effs.count(axis=1) > 1]\n",
    "    vars = vars.loc[effs.index]\n",
    "    alleles = alleles.loc[effs.index]\n",
    "\n",
    "    # Use different combinations of groups\n",
    "    for group, ss_dict in zip(['ALL', 'AFR_EUR', 'EUR'], [pheno_ss, pheno_ss_AFR_EUR, pheno_ss_EUR]):\n",
    "        print('\\t', group)\n",
    "        ss_group = ss_dict[p]\n",
    "        cohorts = [f.split('/')[0] for f in ss_group]\n",
    "        use_effs, use_vars, use_alleles = effs[cohorts], vars[cohorts], alleles[cohorts]\n",
    "\n",
    "        meta_results_fe_rows = []\n",
    "        meta_results_re_rows = []\n",
    "\n",
    "        for variant in use_effs.index:\n",
    "            effects = use_effs.loc[variant].dropna()\n",
    "            variance = use_vars.loc[variant].dropna()\n",
    "            if len(effects) < 2:\n",
    "                continue\n",
    "            v_alleles = use_alleles.loc[variant].dropna()\n",
    "            use_allele = v_alleles.mode().loc[0]\n",
    "            effects.loc[v_alleles != use_allele] = -effects.loc[v_alleles != use_allele]\n",
    "\n",
    "            with warnings.catch_warnings():\n",
    "                warnings.filterwarnings('ignore', r'invalid value encountered in sqrt')\n",
    "                warnings.filterwarnings('ignore', r'divide by zero encountered')\n",
    "                results = combine_effects(effects, variance, method_re='chi2', row_names=effects.index).summary_frame()\n",
    "            results['ID'] = variant\n",
    "            results['N_Studies'] = len(effects)\n",
    "            results['Effect_Allele'] = use_allele\n",
    "\n",
    "            meta_results_re_rows.append(results.loc['random effect', ['Effect_Allele', 'eff', 'sd_eff', 'ci_low', 'ci_upp', 'ID', 'N_Studies']])\n",
    "            meta_results_fe_rows.append(results.loc['fixed effect', ['Effect_Allele', 'eff', 'sd_eff', 'ci_low', 'ci_upp', 'ID', 'N_Studies']])\n",
    "\n",
    "        meta_results_re = pd.concat(meta_results_re_rows, axis=1).transpose().set_index('ID').dropna(subset='sd_eff')\n",
    "        meta_results_fe = pd.concat(meta_results_fe_rows, axis=1).transpose().set_index('ID').dropna(subset='sd_eff')\n",
    "\n",
    "        output_file_re = f'Meta_Output/{p}.{group}.RE.csv.gz'\n",
    "        output_file_fe = f'Meta_Output/{p}.{group}.FE.csv.gz'\n",
    "\n",
    "        meta_results_re.to_csv(output_file_re)\n",
    "        meta_results_fe.to_csv(output_file_fe)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}